### PR TITLE
Adding support for extending with custom queries

### PR DESCRIPTION
**Pull request by pvinis**
### Summary

After the discussion on https://github.com/callstack/react-native-testing-library/pull/569 and https://github.com/callstack/react-native-testing-library/issues/566, I tried to make the prototype for this.

I think it's kind of alright, but I need some help with the flowtyping, as I'm not too familiar with it.

### CODE REVIEW

**Comment by michaelduminy** [comment]
This would be a huge time saver while we wait for other PRs to be merged. +1

**Comment by klekowskim** [comment]
I personally would go with the same API and behaviour as in react-testing-library: https://github.com/testing-library/react-testing-library/blob/c546a6f4927f925bf1187e631ca0444001f067f5/src/pure.js#L35

**Comment by pvinis** [comment]
I can rename it 👍

**Comment by mdjastrzebski** [comment]
I like the idea of allowing custom queries in RNTL.

Big heads up to using the same API as React Testing Library because:
- we re-use RTLs effort that went into creating that API & avoid potential mistakes
- we allow devs to easily transition between web/RN tests
- we allow some code reusage between web/RN

RTL's also exposes (`buildQueries`)[https://github.com/testing-library/dom-testing-library/blob/03cfef8498a5229851984c71c0d6ba8d93ff835d/src/query-helpers.js#L143] so that devs can easily add new quieries in all variants (getBy, getAllBy, queryBy,...) by providing just `queryAllBy` variant. It would be good to provide that as well.

**Comment by thymikee** [comment]
> RTL's also exposes `buildQueries` so that devs can easily add new quieries in all variants (getBy, getAllBy, queryBy,...) by providing just `queryAllBy` variant. It would be good to provide that as well.

Sounds like a good idea for another contribution, that would align with https://github.com/callstack/react-native-testing-library/issues/325

**Comment by pvinis** [comment]
I renamed it. Is this good to go then? And after this is merged we can work on the buildQueries with the refactoring anyway.

**Comment by thymikee** [comment]
Can you make sure lint and flow pass?

**Comment by pvinis** [comment]
Yup yup. I thought the last commit would fix that 😬.

**Comment by pvinis** [comment]
Interestingly it complains even though I have an ignore right above.
https://github.com/callstack/react-native-testing-library/pull/573/files#diff-9c1abad9b45f5a8bc3d32761977018ab6ede546246de46f8c7b1cbf75b320651R384

File: src/render.js
Comment (thymikee) [inline src/render.js:47]: Struggling a bit on how to type it better, so that custom queries are inferred without forced generics. We can improve it later though

**Review by thymikee — APPROVED** [review]
Looks good! I made some final adjustments for the API to be compatible with Testing Library

File: src/__tests__/render.test.js
Comment (mdjastrzebski) [inline src/__tests__/render.test.js:386]:
```suggestion
test('uses custom queries', () => {
```

File: typings/index.d.ts
Comment (mdjastrzebski) [inline typings/index.d.ts:296]: I check RTL/DTL and they seem to have quite decent TS typings for custom queries:

```typescript
export type Query = (
    container: HTMLElement,
    ...args: any[]
) => Error | Promise<HTMLElement[]> | Promise<HTMLElement> | HTMLElement[] | HTMLElement | null;

export interface Queries {
    [T: string]: Query;
}
```
Source: https://github.com/testing-library/dom-testing-library/blob/master/types/get-queries-for-element.d.ts

```typescript
export type RenderResult<Q extends Queries = typeof queries> = {
  // ...
} & {[P in keyof Q]: BoundFunction<Q[P]>}

export interface RenderOptions<Q extends Queries = typeof queries> {
  // ...
  queries?: Q
}
```
Source: https://github.com/testing-library/react-testing-library/blob/master/types/index.d.ts

IMO we could have the same at least for TS, possibly for flow as well if possible.

Comment (mdjastrzebski) [reply typings/index.d.ts:296]: Please note that the `Query` type restricts return value to (Promise)(Array) Element + null + Error. IMO if we provide element queries then we should also consider similar restriction. wdyt?

**Review by mdjastrzebski — COMMENTED** [review]
Overall

**Comment by thymikee** [comment]
@pvinis we just released [v7.1.0](https://github.com/callstack/react-native-testing-library/releases/tag/v7.1.0) with `container` exposed from render, so `queries` is no longer necessary for your use case. We can wait a bit more with merging this unless we have a good type story around that for Flow and TS.

**Comment by mdjastrzebski** [comment]
Closing the PR as a stale code, however the idea is worth pursuing so I've created an issue #971 based on this PR.

**Pull request closed without merging**
